aidial_adapter_openai/gpt4_multi_modal/transformation.py

```diff
@@ -0,0 +1,129 @@
+from typing import List, Optional, cast
+
+from pydantic import BaseModel, root_validator
+
+from aidial_adapter_openai.gpt4_multi_modal.download import (
+    ImageFail,
+    download_image,
+)
+from aidial_adapter_openai.gpt4_multi_modal.image_tokenizer import (
+    tokenize_image,
+)
+from aidial_adapter_openai.gpt4_multi_modal.messages import (
+    create_image_message,
+    create_text_message,
+)
+from aidial_adapter_openai.utils.image_data_url import ImageDataURL
+from aidial_adapter_openai.utils.log_config import logger
+from aidial_adapter_openai.utils.storage import FileStorage
+from aidial_adapter_openai.utils.tokens import Tokenizer
+
+
+class MessageTransformResult(BaseModel):
+    text_tokens: int = 0
+    image_tokens: int = 0
+    errors: Optional[List[ImageFail]] = None
+    message: dict
+
+    @property
+    def total_tokens(self) -> int:
+        return self.text_tokens + self.image_tokens
+
+
+class TransformMessagesResult(BaseModel):
+    error_message: Optional[str] = None
+    transformations: Optional[List[MessageTransformResult]] = None
+
+    @root_validator(pre=True)
+    def validate_info(cls, values):
+        if (
+            values.get("error_message") is None
+            and values.get("transformations") is None
+        ):
+            raise ValueError(
+                "Either 'error_message' or 'transformations' must be provided."
+            )
+        return values
+
+
+class TruncateTransformedMessagesResult(BaseModel):
```

Comment:
Please move to a dedicated module with `truncate_prompt`-related logic.

It's not used in the current module. 
This module is about conversion from GPT4 request to GPT-4o request.